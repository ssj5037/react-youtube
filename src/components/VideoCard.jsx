import React from "react";
import { useNavigate } from "react-router";
import { format } from "timeago.js";

export default function VideoCard({ video }) {
  const navigate = useNavigate();

  return (
    <li
      className="cursor-pointer"
      //   onClick={() => navigate(`/watch/${video.id}`, { state: { video } })}
      onClick={() => navigate(`/watch/${video.id}`)}
    >
      <figure className="overflow-hidden rounded-xl">
        <img
          className="rounded-xl hover:scale-105 transition-all "
          id={video.id}
          width={"100%"}
          height={"100%"}
          src={video.snippet.thumbnails.medium.url}
          alt={video.snippet.title}
        ></img>
      </figure>
      <p className="text-lg line-clamp-2">{video.snippet.title}</p>
      <p className="text-sm text-gray-500">{video.snippet.channelTitle}</p>
      <p className="text-sm text-gray-500">
        {format(new Date(video.snippet.publishedAt).getTime(), "ko")}
      </p>
    </li>
  );
}
