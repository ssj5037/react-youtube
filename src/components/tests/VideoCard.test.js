import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import VideoCard from "../VideoCard";
import { Route } from "react-router-dom";
import { format } from "timeago.js";
import { fakeVideo as video } from "../../tests/videos";
import { withRouter } from "../../tests/utils";
import renderer from "react-test-renderer";

describe("VideoCard", () => {
  const { title, thumbnails, channelTitle, publishedAt } = video.snippet;

  // 정적 테스트 => snapshot
  it("renders grid type correctly", () => {
    const component = renderer
      .create(
        withRouter(<Route path="/" element={<VideoCard video={video} />} />)
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });

  it("renders list type correctly", () => {
    const component = renderer
      .create(
        withRouter(
          <Route
            path="/"
            element={
              <VideoCard
                video={video}
                // type={'list'}
              />
            }
          />
        )
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });

  it("renders video item", () => {
    render(
      withRouter(<Route path="/" element={<VideoCard video={video} />} />)
    );
    const image = screen.getByRole("img");
    expect(image.src).toBe(thumbnails.medium.url);
    expect(image.alt).toBe(title);
    expect(screen.getByText(title)).toBeInTheDocument();
    expect(screen.getByText(channelTitle)).toBeInTheDocument();
    expect(
      screen.getByText(format(new Date(publishedAt).getTime(), "ko"))
    ).toBeInTheDocument();
  });

  // 동적 테스트
  it("navigates to detailed video page with video state when clicked", async () => {
    // 이동하는지 확인하기 위한 테스트용 컴포넌트
    function LocationStateDisplay() {
      return <pre>컴포넌트 이동</pre>;
    }
    render(
      withRouter(
        <>
          <Route path="/" element={<VideoCard video={video} />} />
          <Route
            path={`/watch/${video.id}`}
            element={<LocationStateDisplay />}
          />
        </>
      )
    );

    const card = screen.getByRole("listitem");
    await userEvent.click(card);
    expect(screen.getByText("컴포넌트 이동")).toBeInTheDocument();
  });
});
