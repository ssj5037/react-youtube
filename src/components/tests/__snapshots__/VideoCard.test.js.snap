// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoCard renders grid type correctly 1`] = `
<li
  className="cursor-pointer"
  onClick={[Function]}
>
  <figure
    className="overflow-hidden rounded-xl"
  >
    <img
      alt="title"
      className="rounded-xl hover:scale-105 transition-all "
      height="100%"
      id={1}
      src="http://image/"
      width="100%"
    />
  </figure>
  <p
    className="text-lg line-clamp-2"
  >
    title
  </p>
  <p
    className="text-sm text-gray-500"
  >
    channelTitle
  </p>
  <p
    className="text-sm text-gray-500"
  >
    just now
  </p>
</li>
`;

exports[`VideoCard renders list type correctly 1`] = `
<li
  className="cursor-pointer"
  onClick={[Function]}
>
  <figure
    className="overflow-hidden rounded-xl"
  >
    <img
      alt="title"
      className="rounded-xl hover:scale-105 transition-all "
      height="100%"
      id={1}
      src="http://image/"
      width="100%"
    />
  </figure>
  <p
    className="text-lg line-clamp-2"
  >
    title
  </p>
  <p
    className="text-sm text-gray-500"
  >
    channelTitle
  </p>
  <p
    className="text-sm text-gray-500"
  >
    just now
  </p>
</li>
`;
